### PR TITLE
cephfs-t2-CEPH-83574107-and-utils-fix-to-handle-multifs

### PR DESCRIPTION
**Pull request by hkadam134**
Tier2 - Automation - CEPH-83574107 - Validate cephfs-mirroring on multifs setup
Also has a few fixes to cephfs-utils file to handle multiple filesystem scenarios.
Logs : http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-4K11X8/

# Description

Please include Automation development guidelines. Source of Test case - New Feature/Regression Test/Close loop of customer BZs
<details>

<summary>click to expand checklist</summary>

- [ ] Create a test case in Polarion reviewed and approved.
- [ ] Create a design/automation approach doc. Optional for tests with similar tests already automated.
- [ ] Review the automation design
- [ ] Implement the test script and perform test runs
- [ ] Submit PR for code review and approve
- [ ] Update Polarion Test with Automation script details and update automation fields
- [ ] If automation is part of Close loop, update BZ flag qe-test_coverage “+” and link Polarion test
</details>

### CODE REVIEW

**Comment by hkadam134** [comment]
Entire Suite Log : http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-CWR7KA/

File: tests/cephfs/cephfs_mirroring/test_cephfs_mirroring_configure_cephfs_mirroring.py [hunk 1]
```diff
@@ -53,8 +55,69 @@ def run(ceph_cluster, **kw):
...
+        for fs in result:
+            fs_name = fs["name"]
+            target_clients[0].exec_command(
+                sudo=True, cmd=f"ceph fs volume rm {fs_name} --yes-i-really-mean-it"
+            )
+        time.sleep(30)
+
+        mds_nodes = ceph_cluster_dict.get("ceph1").get_ceph_objects("mds")
+        log.info(f"Available MDS Nodes {mds_nodes[0]}")
+        print(len(mds_nodes))
```
Comment (neha-gangadhar) [inline tests/cephfs/cephfs_mirroring/test_cephfs_mirroring_configure_cephfs_mirroring.py:91]: please modify print() with loggers wherever used in the file.

File: tests/cephfs/cephfs_mirroring/cephfs_mirroring_utils.py [hunk 2]
```diff
@@ -386,10 +386,25 @@ def validate_synchronization(
...
+            log.info(f"File ID is {filesystem_id}")
+            name = filesystem["name"]
+            log.info(f"name  is {name}")
+            peers = filesystem["peers"]
+            log.info(f"peers is {peers}")
+            for peer in peers:
+                peer_uuid = peer["uuid"]
+                log.info(f"peer uuid is {peer_uuid}")
+                if name == fs_name:
+                    filesystem_info[filesystem_id] = peer_uuid
```
Comment (AmarnatReddy) [inline tests/cephfs/cephfs_mirroring/cephfs_mirroring_utils.py:402]: can you make get_peer_uuid as seperate function by accepting filesystem name as argument
get_peer_uuid(fs_name="cephfs")

File: tests/cephfs/cephfs_mirroring/cephfs_mirroring_utils.py [hunk 3]
```diff
@@ -451,31 +467,30 @@ def validate_snapshot_sync_status(
...
+                    snaps_synced = status.get("snaps_synced")
+                    return {
+                        "snapshot_name": snapshot_name,
+                        "sync_duration": sync_duration,
+                        "sync_time_stamp": sync_time_stamp,
+                        "snaps_synced": snaps_synced,
+                    }
+                    log.info("ALl snapshots are synced")
+        else:
+            log.error(f"{snapshot_name} not synced")
```
Comment (AmarnatReddy) [inline tests/cephfs/cephfs_mirroring/cephfs_mirroring_utils.py:492]: Can you add last sync details also in the error log 
log.error(f"{snapshot_name} not synced") --> log.error(f"{snapshot_name} not synced, \n last synced data is {data}")

File: tests/cephfs/cephfs_mirroring/cephfs_mirroring_utils.py (hunk 2 above)
Comment (hkadam134) [reply tests/cephfs/cephfs_mirroring/cephfs_mirroring_utils.py:402]: `log.info("Fetch the FSID of the ceph cluster")
def get_fsid(self, cephfs_mirror_node):
    out, _ = cephfs_mirror_node.exec_command(sudo=True, cmd="cephadm ls")
    data = json.loads(out)
    fsid = data[0]["fsid"]
    return fsid


log.info("Fetch the cephfs-mirror daemon name")
def get_daemon_name(self, source_clients):
    out, _ = source_clients.exec_command(sudo=True, cmd="ceph orch ps --daemon_type cephfs-mirror -f json")
    data = json.loads(out)
    daemon_name = data[0]["daemon_name"]
    return daemon_name


log.info("Fetch the Filesystem Info to identify the Peer ID of the Peer Connection")
def get_filesystem_info_for_fetching_peer_id(self, source_clients, fs_name):
    out, _ = source_clients.exec_command(sudo=True, cmd="ceph fs snapshot mirror daemon status -f json")
    data = json.loads(out)
    filesystem_info = {}
    for filesystem in data[0]["filesystems"]:
        filesystem_id = filesystem["filesystem_id"]
        name = filesystem["name"]
        peers = filesystem["peers"]
        for peer in peers:
            peer_uuid = peer["uuid"]
            if name == fs_name:
                filesystem_info[filesystem_id] = peer_uuid
    return filesystem_info

def get_asok_file(self, cephfs_mirror_node, fsid, daemon_name):
    cmd = f"cd /var/run/ceph/{fsid}/ ; ls -1tr ceph-client.{daemon_name}* | head -n 1 | tr -d '\\n'"
    file = cephfs_mirror_node.exec_command(sudo=True, cmd=cmd)
    asok_file = file[0].replace("\\n", "")
    return asok_file

`

Comment (hkadam134) [reply tests/cephfs/cephfs_mirroring/cephfs_mirroring_utils.py:402]: @AmarnatReddy = Shall I split like this ?

Comment (hkadam134) [reply tests/cephfs/cephfs_mirroring/cephfs_mirroring_utils.py:402]: def validate_synchronization(self, cephfs_mirror_node, source_clients, fs_name, snap_count):
    fsid = self.get_fsid(cephfs_mirror_node)
    daemon_name = self.get_daemon_name(source_clients)
    filesystem_info = self.get_filesystem_info(source_clients, fs_name)
    for filesystem_id, peer_uuid in filesystem_info.items():
    asok_file = self.get_asok_file(cephfs_mirror_node, fsid, daemon_name)
    
    return fsid, asok_file, filesystem_id, peer_uuid

File: tests/cephfs/cephfs_mirroring/test_cephfs_mirroring_validate_cephfs-mirroring_on_multifs_setup.py [hunk 4]
```diff
@@ -0,0 +1,575 @@
...
+        target_clients[0].exec_command(
+            sudo=True, cmd=f'ceph fs volume create {target_fs_2} --placement="2 {mds_hosts_2}"'
+        )
+        fs_util_ceph1.wait_for_mds_process(target_clients[0], target_fs_2)
+
+        target_user1 = "mirror_remote1"
+        target_user2 = "mirror_remote2"
+        target_site_name = "remote_site"
+
+        log.info("Enable mirroring mgr module on source and destination")
```
Comment (AmarnatReddy) [inline tests/cephfs/cephfs_mirroring/test_cephfs_mirroring_validate_cephfs-mirroring_on_multifs_setup.py:139]: Can we use for loop for whole testing hemanth. As whole process is same .
we are calling same functions twice

File: tests/cephfs/cephfs_mirroring/cephfs_mirroring_utils.py [hunk 5]
```diff
@@ -541,17 +597,16 @@ def list_and_verify_remote_snapshots_and_data(
             tuple: A tuple containing a boolean indicating success or failure, and a message.
         """
         try:
-            target_clients.exec_command(sudo=True, cmd=f"mkdir {target_mount_path}")
+            target_clients.exec_command(sudo=True, cmd=f"mkdir -p {target_mount_path}")
```
Comment (AmarnatReddy) [inline tests/cephfs/cephfs_mirroring/cephfs_mirroring_utils.py:600]: can we add finally block for removing target_mount_path
```
finally:
            target_clients.exec_command(sudo=True, cmd=f"umount {target_mount_path}")
            target_clients.exec_command(sudo=True, cmd=f"rm -rf {target_mount_path}")
```

**Review by AmarnatReddy — APPROVED** [review]
LGTM

**Pull request merged**